src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitTest.java

```diff
@@ -72,6 +139,79 @@ public void nonProtectedCheckMethodsTest() {
                 .doNotHaveModifier(JavaModifier.ABSTRACT)
                 .should().notBeProtected();
 
-        checkMethodsShouldNotBeProtectedRule.check(importedClasses);
+        checkMethodsShouldNotBeProtectedRule.check(CHECKS_PACKAGE);
+    }
+
+    /**
+     * Tests that all checks either extend {@link AbstractCheck} or {@link AbstractFileSetCheck}.
+     *
+     * @noinspection JUnitTestMethodWithNoAssertions
+     * @noinspectionreason JUnitTestMethodWithNoAssertions - asserts in callstack,
+     *     but not in this method
+     */
+    @Test
+    public void testChecksShouldExtendAbstractCheckOrAbstractFileSetCheck() {
+        final ArchCondition<JavaClass> beSuppressedClass = new SuppressionArchCondition<>(
+            SUPPRESSED_CLASSES, "be suppressed");
+
+        final ArchRule checksShouldExtendCorrectAbstractClass = classes()
+            .that()
+            .haveSimpleNameEndingWith("Check")
+            .should(extend(AbstractCheck.class))
```

Comment:
Used `beSubclassOf(..)`

Reply:
subclass means all hiearchy of classes.
"Super" is most precise term here

Reply:
Can we use `beDirectSubclassOf(..)`? otherwise, we would have the problem of has/have, ArchUnit also doesn't use has/have analogy. Classes in checks package should has superclass is not the correct usage.

Reply:
>we would have the problem of has/hav

What is a problem ?
In our context it is `has`

Reply:
https://www.google.com/amp/s/textranch.com/27619/it-should-have/or/it-should-has/amp/

Reply:
It would be an ArchCondition for a Java Class which is singular, but not big of an issue, we can go with `have`.